### PR TITLE
Support for commands output formatting with templates: initial version

### PR DESCRIPTION
**Pull request by clockworksoul**
Support for output formatting templates.

This will resolve https://github.com/getgort/gort/issues/116. The changes are BIG.

TL;DR:

- Templates can be set at the config, bundle, and command level. There are also default templates that work fine.
- The template syntax is based on Go templates. For example, this is the default command error message template:

```
{{ header | color "#FF0000" }}{{ .Response.Title }}{{ endheader }}
{{ text }}The pipeline failed planning the invocation:{{ endtext }}
{{ text | monospace true }}{{ .Request.Bundle.Name }}:{{ .Request.Command.Name }} {{ .Request.Parameters }}{{ endtext }}
{{ text }}The specific error was:{{ endtext }}
{{ text | monospace true }}{{ .Response.Out }}{{ endtext }}
```

- Message elements are described using custom Go template functions (`header`, `text`, etc)
- On command execution, the resulting `CommandResponseEnvelope` value is passed to the templating engine.
  - The adapter will detect whether command output is JSON-formatted (json.Valid does this easily)
    - If yes: the response if unmarshalled into an `interface{}` value (envelope.Payload) where it is accessible via Go template syntax
    - If not: the response text is placed, unmodified, into the envelope’s `Payload` as a string.
- Moved some repeated and more general logic into the `adapter` package and out of the implementations.

I'll write more later.

TODO still:
* Documentation (will come in a separate PR).

Open Question:
* Should a given template be expected to work for ALL chat providers, or should we support the ability to set templates at the chat level?
  * This would mean that each adapter in the config would support template definitions (easy)
  * Each command and bundle would also need support for setting each of the two command template types for each supported adapter type. Is that too much complexity?

### CODE REVIEW

File: data/bundle.go [hunk 1]
```diff
@@ -45,6 +46,18 @@ type Bundle struct {
 	Permissions       []string                  `yaml:",omitempty" json:"permissions,omitempty"`
 	Commands          map[string]*BundleCommand `yaml:",omitempty" json:"commands,omitempty"`
 	Default           bool                      `yaml:"-" json:"default,omitempty"`
+	Templates         Templates                 `yaml:",omitempty" json:"templates,omitempty"`
+}
+
+// BundleCommand represents a bundle command, as defined in the "bundles/commands"
```
Comment (clockworksoul) [inline data/bundle.go:52]: This was moved, and a `Templates` value added.

**Comment by theothertomelliott** [comment]
Love the idea of using templates for flexibility!

I feel that it would be nice if we could have one model for all chat platforms, maybe even if they don't all support every tag or function. Anything to make it easier to create community-built bundles will be a win if you ask me.

**Comment by clockworksoul** [comment]
> Love the idea of using templates for flexibility!
> 
> I feel that it would be nice if we could have one model for all chat platforms, maybe even if they don't all support every tag or function. Anything to make it easier to create community-built bundles will be a win if you ask me.

Absolutely agree! That's the goal for sure. It's complicated by the fact that each chat implementation seems to have a wildly different window model, but I think we can get pretty close.

I'm wondering though whether we might (at some point) need to allow users to override those the general templates in some cases because a particular implementation simply refuses to render correctly. We don't need to do that now, but at some point I'm afraid that may become an issue.

**Comment by theothertomelliott** [comment]
> I'm wondering though whether we might (at some point) need to allow users to override those the general templates in some cases because a particular implementation simply refuses to render correctly. We don't need to do that now, but at some point I'm afraid that may become an issue.

Maybe we should have a chat about some specific use cases we'd like to target, give us some guidance about the kinds of formatting we want to support as first-class.

**Comment by clockworksoul** [comment]
> Maybe we should have a chat about some specific use cases we'd like to target, give us some guidance about the kinds of formatting we want to support as first-class.

Absolutely!

**Review by theothertomelliott — APPROVED** [review]
